packages/engineer/feature/dev-server.dev-server.env.ts

```diff
@@ -0,0 +1,227 @@
+                })
+            );
+
+            disposables.add(() => application.getNodeEnvManager()?.closeAll());
+
+            if (engineConfig && engineConfig.serveStatic) {
+                for (const { route, directoryPath } of engineConfig.serveStatic) {
+                    app.use(route, express.static(directoryPath));
+                }
+            }
+
+            const topologyOverrides = (featureName: string): Record<string, string> | undefined =>
+                featureName.indexOf('engineer/') === 0
+                    ? {
+                          [devServerEnv.env]: `http://localhost:${port}/${devServerEnv.env}`,
+                      }
+                    : undefined;
+
+            app.use(`/${publicConfigsRoute}`, [
+                // WTF need to look into
+                ensureTopLevelConfigMiddleware,
+                // I think this is irrelevant, apps should handle their own communication, I shouldn't do it for them
+                createCommunicationMiddleware(application.getNodeEnvManager()!, publicPath, topologyOverrides),
+                // WTF?
+                createLiveConfigsMiddleware(configurations, basePath, application.getOverrideConfigsMap()),
+
+                // WTF?
+                createConfigMiddleware(overrideConfig),
+            ]);
+
+            // Write middleware for each of the apps
+            for (const childCompiler of compiler.compilers) {
+                const devMiddleware = WebpackDevMiddleware(childCompiler, {
+                    publicPath: '/',
+                    logLevel: 'silent',
+                });
+                disposables.add(() => new Promise((res) => devMiddleware.close(res)));
+                app.use(devMiddleware);
+            }
+
+            // Why would I run this script when the compiler is done?
+            await new Promise((resolve) => {
+                compiler.hooks.done.tap('engine-scripts init', resolve);
+            });
+
+            const featureEnvDefinitions = application.getFeatureEnvDefinitions(features, configurations);
+
+            app.use(
+                '/engine-feature',
+                createFeaturesEngineRouter(application.getOverrideConfigsMap(), application.getNodeEnvManager()!)
+            );
+
+            app.get('/engine-state', (_req, res) => {
+                res.json({
+                    result: 'success',
+                    data: {
+                        features: featureEnvDefinitions,
+                        featuresWithRunningNodeEnvs: application
+                            .getNodeEnvManager()
+                            ?.getFeaturesWithRunningEnvironments(),
+                    },
+                });
+            });
+
+            if (autoLaunch && featureName) {
+                await application.runFeature({
+                    featureName,
+                    configName,
+                });
+            }
+
+            const engineerCompilers = webpack([...engineerWebpackConfigs]);
```

Comment:
Added a comment, not sure about the extracting to a function, since the fact that both compilers do the same thing is by accident, they aren't really coupled to each other

Reply:
just a function that does that, for all the logic to not be in the run function itself.
call it runEngineerWabpackCompiler